metricbeat/module/kafka/partition/partition.go

```diff
@@ -19,75 +21,196 @@ func init() {
 // MetricSet type defines all fields of the partition MetricSet
 type MetricSet struct {
 	mb.BaseMetricSet
-	client sarama.Client
+
+	broker *sarama.Broker
+	cfg    *sarama.Config
+	id     int32
 }
 
-// New creates a new instance of the partition MetricSet
+var noID int32 = -1
+
+var errFailQueryOffset = errors.New("Failed to query offset for")
+
+// New create a new instance of the partition MetricSet
 func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
-	logp.Warn("EXPERIMENTAL: The %v %v metricset is experimental", base.Module().Name(), base.Name())
+	config := struct{}{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	cfg := sarama.NewConfig()
+	cfg.Net.DialTimeout = base.Module().Config().Timeout
+	cfg.Net.ReadTimeout = base.Module().Config().Timeout
+	cfg.ClientID = "metricbeat"
+
+	broker := sarama.NewBroker(base.Host())
+	return &MetricSet{
+		BaseMetricSet: base,
+		broker:        broker,
+		cfg:           cfg,
+		id:            noID,
+	}, nil
+}
+
+func (m *MetricSet) connect() (*sarama.Broker, error) {
+	b := m.broker
+	if err := b.Open(m.cfg); err != nil {
+		return nil, err
+	}
+
+	if m.id != noID {
+		return b, nil
+	}
 
-	return &MetricSet{BaseMetricSet: base}, nil
+	// current broker is bootstrap only. Get metadata to find id:
+	meta, err := b.GetMetadata(&sarama.MetadataRequest{})
+	if err != nil {
+		return nil, err
+	}
```

Comment:
See default sarama settings below. As we do the connect for each call I think it can be ok if one of them fails.